project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module

```diff
@@ -185,10 +185,261 @@ function c4m_og_menu() {
+  $form['gid'] = array(
+    '#type' => 'value',
+    '#value' => $node->nid,
+  );
+
+  $form['#submit'][] = 'c4m_og_approve_group_request_submit';
+  $form['#submit'][] = 'c4m_user_notifications_notify_user_group_approved';
+  $form['#submit'][] = 'c4m_user_notifications_log_group_approved';
+  $form['#submit'][] = 'c4m_og_route_to_group';
+
+  return confirm_form($form, t('Please enter your comments:'), 'node/' . $node->nid . '/group', '', t('Approve'), t('Cancel'));
+}
+
+/**
+ * Function initializes form with input values, and returns confirm form.
+ *
+ * @param object $form
+ *   Form array to populate.
+ * @param object $form_state
+ *   Form state array.
+ * @param object $node
+ *   Group node.
+ *
+ * @return object
+ *   Approve group request form.
+ */
+function c4m_og_reject_group_request($form, &$form_state, $node) {
+  if (!_c4m_og_group_at_initial_state($node->nid)) {
+    // If group is not at initial state, present form with failure reason.
+    $message = _c4m_og_format_group_request_failure_reason_message($node->nid);
+    drupal_set_message($message, 'warning');
+
+    return $form;
+  }
+
+  $form['message_from_admin'] = array(
+    '#type' => 'textarea',
+  );
+
+  $form['gid'] = array(
+    '#type' => 'value',
+    '#value' => $node->nid,
+  );
+
+  $form['#submit'][] = 'c4m_og_reject_group_request_submit';
+  $form['#submit'][] = 'c4m_user_notifications_log_group_rejected';
+  $form['#submit'][] = 'c4m_og_route_to_group';
+
+  return confirm_form($form, t('Please enter a reason:'), 'node/' . $node->nid . '/group', '', t('Reject'), t('Cancel'));
+}
+
+/**
+ * Submits approve group request form.
+ *
+ * Gets inputs from form, and activates c4m_og_control_group_request() function
+ * with action = 'approve'.
+ *
+ * @param object $form
+ *   Form array to populate.
+ * @param object $form_state
+ *   Form state array.
+ */
+function c4m_og_approve_group_request_submit($form, &$form_state) {
+  $gid = $form_state['values']['gid'];
+
+  c4m_og_control_group_request('approve', $gid);
+}
+
+/**
+ * Submits reject group request form.
+ *
+ * Gets inputs from form, and activates c4m_og_control_group_request() function
+ * with action = 'reject'.
+ *
+ * @param object $form
+ *   Form array to populate.
+ * @param object $form_state
+ *   Form state array.
+ */
+function c4m_og_reject_group_request_submit($form, &$form_state) {
+  $gid = $form_state['values']['gid'];
+
+  c4m_og_control_group_request('reject', $gid);
+}
+
+/**
+ * Changes group state to approved / rejected (per $action).
+ *
+ * @param string $action
+ *   Use 'approve' when group is approved.
+ *   Use 'reject' when group is rejected.
+ * @param int $gid
+ *   Group ID.
+ */
+function c4m_og_control_group_request($action, $gid) {
+  // Check that group is at initial ('pending') state.
+  $wrapper = entity_metadata_wrapper('node', $gid);
+
+  if (!_c4m_og_group_at_initial_state($gid)) {
+    // If group is not at initial state, present warning message and return.
```

Comment:
replace the technical comment with a simple one, something like "request already handled ..."